### ISSUE TITLE
"Suboptimal present of frame i" warn log spam on Android

### ISSUE THREAD

**Issue by James2022-rgb** [gfx-rs/wgpu#3345]
## Description

`log::warn!("Suboptimal present of frame {}", texture.index);` warn log may be emitted on every call to `vkQueuePresentKHR` via `wgpu::SurfaceTexture::present`, depending on the device and its current orientation.

Log output example from a session on my Galaxy S20 Ultra 5G (Adreno 650), Android 10:
```
01-03 11:25:00.523  3115  5742 W wgpu_hal::vulkan: [C:\Users\James\.cargo\registry\src\github.com-1ecc6299db9ec823\wgpu-hal-0.14.1\src\vulkan\mod.rs:659] Suboptimal present of frame 0
01-03 11:25:00.539  3115  5742 W wgpu_hal::vulkan: [C:\Users\James\.cargo\registry\src\github.com-1ecc6299db9ec823\wgpu-hal-0.14.1\src\vulkan\mod.rs:659] Suboptimal present of frame 1
01-03 11:25:00.556  3115  5742 W wgpu_hal::vulkan: [C:\Users\James\.cargo\registry\src\github.com-1ecc6299db9ec823\wgpu-hal-0.14.1\src\vulkan\mod.rs:659] Suboptimal present of frame 2
01-03 11:25:00.572  3115  5742 W wgpu_hal::vulkan: [C:\Users\James\.cargo\registry\src\github.com-1ecc6299db9ec823\wgpu-hal-0.14.1\src\vulkan\mod.rs:659] Suboptimal present of frame 0
01-03 11:25:00.590  3115  5742 W wgpu_hal::vulkan: [C:\Users\James\.cargo\registry\src\github.com-1ecc6299db9ec823\wgpu-hal-0.14.1\src\vulkan\mod.rs:659] Suboptimal present of frame 1
01-03 11:25:00.606  3115  5742 W wgpu_hal::vulkan: [C:\Users\James\.cargo\registry\src\github.com-1ecc6299db9ec823\wgpu-hal-0.14.1\src\vulkan\mod.rs:659] Suboptimal present of frame 2
...
...
```

Relevant piece of code in _wgpu-hal_:
https://github.com/gfx-rs/wgpu/blob/186a29c34d54d8628bbebb998a537210ac565b71/wgpu-hal/src/vulkan/mod.rs#L650-L662

It is highly likely that this log spamming occurs when the Android device's current orientation is anything other than `VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR`,
thus requiring final surface rotation on Android compositor's (SurfaceFlinger) side.
(Which may have to preempt the app's GPU work, incurring performance hit)

## Direct cause

Starting from Android 10 (to my knowledge), Android libvulkan's implementation of `vkQueuePresentKHR` returns `VK_SUBOPTIMAL_KHR` if the Vulkan app doesn't do pre-rotation [^pre-rotation],
i.e `VkSwapchainCreateInfoKHR::preTransform` != `VkSurfaceCapabilitiesKHR::currentTransform`.
This is indeed what _wgpu-hal_ does; it just uses `VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR`:
https://github.com/gfx-rs/wgpu/blob/186a29c34d54d8628bbebb998a537210ac565b71/wgpu-hal/src/vulkan/device.rs#L562

Relevant piece of code in Google's Android libvulkan implementation:
https://android.googlesource.com/platform/frameworks/native/+/master/vulkan/libvulkan/swapchain.cpp#2021
```c
VkResult QueuePresentKHR(VkQueue queue, const VkPresentInfoKHR* present_info) {
// --snip--
            if (swapchain.pre_transform != window_transform_hint) {
                swapchain_result =
                    WorstPresentResult(swapchain_result, VK_SUBOPTIMAL_KHR);
            }
// --snip--
```

A tweet thread from Hans-Kristian Arntzen[^tweet] pertains to using `VK_SUBOPTIMAL_KHR` for a control flow for swapchain recreation, but is nonetheless insightful.

## Repro steps

Any Android app that uses _wgpu_ with its Vulkan backend _should_ exhibit the problem if its `android:screenOrientation` value allows the orientation to be anything other than the identity orientation for that device.
Make sure the logging output correctly goes to LogCat.

I could try putting together a quick-and-dirty repro app, if necessary.

# Expected vs observed behavior

Ideally Android pre-rotation would be properly supported, but this would clearly introduce several limitations on the user's side.
Namely having to adjust, according to current device orientation:
- View projection matrix.
- Viewport/scissor region.
- `dFdx`/`dFdy` usage in fragment shaders.

Although this is only a problem when rendering directly to the swapchain image, these limitations seem far from acceptable for a library like _wgpu_.

More realistic solutions/workarounds that come to mind:
- Treat `vkQueuePresentKHR`'s `VK_SUBOPTIMAL_KHR` as `VK_SUCCESS` if `target_os = "android"`, as Hans-Kristian mentions in aforementioned tweet ?
- Change the log level for this particular log to something other than `Warn`, maybe gating with `target_os = "android"` ?
  This log essentially gets emitted every frame when not in identity orientation.

## Platform
Android (_should_ only affect 10 and above), Vulkan backend.

[^pre-rotation]: https://developer.android.com/games/optimize/vulkan-prerotation#swapchain_recreation
[^tweet]: https://twitter.com/Themaister/status/1207062674011574273

**Comment by teoxoy** [comment]
I stumbled across https://developer.android.com/games/optimize/vulkan-prerotation#detect_device_orientation_changes

> The most reliable way to detect an orientation change in your application is to verify whether the `vkQueuePresentKHR()` function returns `VK_SUBOPTIMAL_KHR`.

I think it's safe to assume that on android, `VK_SUBOPTIMAL_KHR` is only returned when the orientation changes.

@James2022-rgb I think a PR for this is welcome!

**Comment by James2022-rgb** [comment]
Resolved with #3525.
